google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/SpannerStatementParser.java

```diff
@@ -250,4 +271,65 @@ ParametersInfo convertPositionalParametersToNamedParametersInternal(char paramCh
     }
     return new ParametersInfo(paramIndex - 1, named.toString());
   }
+
+  private boolean isReturning(String sql, int index) {
+    return (index >= 1)
+        && (index + 12 <= sql.length())
+        && THEN_RETURN_PATTERN.matcher(sql.substring(index - 1, index + 12)).matches();
+  }
+
+  @InternalApi
+  @Override
+  boolean checkReturningClauseInternal(String rawSql) {
```

Comment:
Could we also do a couple of optimizations here:
1. Do a `sql.contains("return") && sql.contains("then")` check before anything else, as that would allow us to return false quickly in most cases.
2. (Maybe in a separate PR, as this is more complicated here because of the checking for escaping etc.) Loop in the opposite direction if the quick check succeeds?

Reply:
Good point! I'll work on enhancement #2 in a separate PR, and will incorporate enhancement #1 in this PR for now.

Reply:
The check can be `sql.contains("then return")` as we must have already replaced all consecutive whitespaces with a single space.

Reply:
No, I would recommend that we do not do the last. So the order of operations would roughly be:
1. sql.toLowerCase()
2. sql.contains("return") && sql.contains("then")
3. sql.replaceAll("\\s+", " ")

The reason for this is that the `replaceAll` is also quite expensive, as it is also a regular expression check for each element in the string.

Reply:
Makes sense.

Reply:
Made this change.